### PR TITLE
add environment design

### PR DESCRIPTION
**Pull request by hongchaodeng**
ref: https://github.com/oam-dev/kubevela/pull/1596, https://github.com/oam-dev/kubevela/pull/999, https://github.com/oam-dev/kubevela/issues/792.

This is the latest effort and thoughts to support Environment as shared base for application deployment.

### CODE REVIEW

File: design/vela-core/environment.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  environment: prod
+```
+
+### 2. EnvConfigDefinition CRD
+
+An Environment can have multiple EnvConfigs. The schema is defined in EnvConfigDefinition CRD:
+
+```yaml
+apiVersion: core.oam.dev/v1beta1
+kind: EnvConfigDefinition
```
Comment (wonderflow) [inline design/vela-core/environment.md:87]: Can we merge it with PolicyDefinition/WorkflowStepDefinition and call them Definition? We should reduce CRD to reduce complexity.

Comment (hongchaodeng) [reply design/vela-core/environment.md:87]: What's your suggestion? How can we do that if they are for different things and have different spec fields?

File: design/vela-core/environment.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# Managing Environments As Shared Bases For Application Deployment
+
+## Background
+
+A team of application development usually needs to prepare some shared environments for developers to deploy applications.
+For example, most teams will have two environments, a *dev* environment for testing their applications in live instances,
```
Comment (zzxwill) [inline design/vela-core/environment.md:6]: Normally we call it `testing` or `staging` environment.

File: design/vela-core/environment.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  name: prod
+spec:
+  # A list of pluggable EnvConfig resources.
+  # The resources will be parsed and applied when Environment CR is appled.
+  configs:
+    - type: cluster
+      properties:
+        name: prod-cluster
+        selector:
+          tier: prod
```
Comment (zzxwill) [inline design/vela-core/environment.md:42]: In `Env`, What we talked about most is `cluster`. I am curious about what's the definition of `cluster`. Could you please set an example for type `cluster`:)

File: design/vela-core/environment.md (hunk 1 above)
Comment (wonderflow) [reply design/vela-core/environment.md:87]: Let's make a unified Object called CUEPackage, and the policy,workflow and env all rely on the CUEPackage as abstraction. And in the implementation, they're all CUE based templates.

Comment (hongchaodeng) [reply design/vela-core/environment.md:87]: First of all, I don't think we should only make policy, workflow, and env. If we want to make a unified object, all definitions should be there, including component and trait. We shouldn't treat policy, etc. as separate definitions.

Second, what about the specific fields that policy, workflow, or env have? They need to be in spec.

File: design/vela-core/environment.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+```yaml
+kind: Environment
+metadata:
+  name: prod
+spec:
+  # A list of pluggable EnvConfig resources.
+  # The resources will be parsed and applied when Environment CR is appled.
+  configs:
+    - type: cluster
```
Comment (wonderflow) [inline design/vela-core/environment.md:38]: please describe the detail of cluster type.

File: design/vela-core/environment.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+```
+
+We also need to add an optional field `environment` in Application:
+
+```yaml
+kind: Application
+spec:
+  # If set, app controller will add the AppRevision into the Environment
+  # CR spec after generating AppRevision before executing Workflow.
+  environment: prod
```
Comment (wonderflow) [inline design/vela-core/environment.md:78]: what if we want to deploy application to more than one environment?

File: design/vela-core/environment.md (hunk 3 above)
Comment (wonderflow) [reply design/vela-core/environment.md:42]: +1

And what if it's not a K8s cluster? It's a env pointing to a cloud?

**Comment by hongchaodeng** [comment]
Closing this for now.
After offline discussion, we probably do not need a new Environment API. Instead I will propose some mechanisms and implementation plans to provide environment management experience.

**Pull request closed without merging**